charms/knative-serving/src/lightkube_helpers/__init__.py

```diff
@@ -0,0 +1,88 @@
+from typing import List, Union, Tuple
+
+import lightkube
+from lightkube.core.resource import NamespacedResource, GlobalResource
+from lightkube.resources.apps_v1 import StatefulSet
+from ops.model import ActiveStatus, WaitingStatus, BlockedStatus, MaintenanceStatus
+
+from charm_helpers.exceptions import ErrorWithStatus
+
+
+def check_resources(client: lightkube.Client, expected_resources: List[Union[NamespacedResource, GlobalResource]]) -> \
+        (bool, List[Exception]):
+    """Checks status of resources in cluster, returning True if all are considered ready
+
+    Note: This is a basic skeleton of a true check on the resources.  Currently it only checks
+    that resources exist and that StatefulSets have their desired number of replicas ready.
+
+    Returns: Tuple of:
+        Status (bool)
+        List of Exceptions encountered during failed checks, with each entry
+        indexed the same as the corresponding expected_resource (list[str])
+    """
+    errors = [None] * len(expected_resources)
+    for i, expected_resource in enumerate(expected_resources):
+        try:
+            found_resource = client.get(
+                type(expected_resource),
+                expected_resource.metadata.name,
+                namespace=expected_resource.metadata.namespace,
+            )
+        except lightkube.core.exceptions.ApiError as e:
+            msg = f"Cannot find k8s object corresponding to '{expected_resource.metadata}'"
+            errors[i] = ResourceNotFoundError(msg, BlockedStatus)
+            continue
+
+        if isinstance(found_resource, StatefulSet):
+            try:
+                validate_statefulset(found_resource)
+            except ReplicasNotReadyError as e:
+                errors[i] = e
+
+    return not any(errors), errors
+
+
+def validate_statefulset(resource: StatefulSet) -> bool:
+    """Returns True if the StatefulSet is ready, else raises an Exception"""
+    readyReplicas = resource.status.ready_replicas
+    replicas_expected = resource.spec.replicas
+    if readyReplicas != replicas_expected:
+        error_message = (
+            f"StatefulSet {resource.metadata.name} in namespace "
+            f"{resource.metadata.namespace} has {readyReplicas} readyReplicas, "
+            f"expected {replicas_expected}"
+        )
+        raise ReplicasNotReadyError(error_message, WaitingStatus)
+
+    return True
+
+
+def get_first_worst_error(errors: List[Exception]) -> Exception:
+    """Returns the first of the worst errors in the list, ranked by their status
+
+    Raises if List contains no Exceptions, or if any Exception does not have a .status
+
+    Status are ranked, starting with the worst:
+        BlockedStatus
+        WaitingStatus
```

Comment:
Given that we have custom exceptions we maybe could do some more granular ranking.

Reply:
Yeah definitely.  I can't think of a case where it matters, but we definitely could.  I'm not sure what to do to make the order flexible... maybe an optional argument?

Could also have a similar helper that concatenates all errors into a single error/log message.  So rather than the first-worst, it maybe could summarize into `BlockedStatus("Encountered 3 blocking events and 2 waiting events")` and then `self.log.info(message_that_lists_all_3_blocks_and_2_waits_in_depth)`